src/Workspaces/Core/Portable/Shared/Extensions/SourceTextExtensions.cs

```diff
@@ -17,15 +19,18 @@ namespace Microsoft.CodeAnalysis.Shared.Extensions;
 
 internal static partial class SourceTextExtensions
 {
-    // char pooled memory : 8K * 256 = 2MB
-    private const int CharArrayLength = 4 * 1024;
+    // char segment: 4k characters. 4K * 256 * 2 (bytes per char) = 4MB
+    private const int CharSegmentLength = 4096;
+
+    // 16k characters. Equivalent to 32KB in memory. comes from SourceText char buffer size and less than large object size
+    public const int SourceTextLengthThreshold = 32 * 1024 / sizeof(char);
```

Comment:
note: we can double this safely.  If you're less than 32k chars you won't be going in the LOH either (since that triggers at 70k bytes).